File: backend/report.py
import os
import random
import sqlite3
import requests
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from datetime import datetime, timedelta
from dotenv import load_dotenv
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, PageBreak
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib.pagesizes import A4
from database import get_connection, get_mood_data

load_dotenv()
GROQ_API_KEY = os.getenv("GROQ_API_KEY")


# ─────────────────────────────────────────────
# Personality type descriptions (MBTI-style)
# ─────────────────────────────────────────────
PERSONALITY_DESCRIPTIONS = {
    "INTJ": {
        "title": "The Architect",
        "summary": "Strategic, independent, and highly analytical. INTJs are driven by logic and long-term vision, preferring depth over breadth in all pursuits.",
        "strengths": "Strategic thinking, decisiveness, high standards",
        "growth": "Opening up emotionally and embracing flexibility",
    },
    "INTP": {
        "title": "The Logician",
        "summary": "Curious and inventive thinkers who love exploring abstract ideas. INTPs seek precision and often lose themselves in thought.",
        "strengths": "Analytical ability, creativity, open-mindedness",
        "growth": "Following through on ideas and social engagement",
    },
    "ENTJ": {
        "title": "The Commander",
        "summary": "Bold, imaginative, and strong-willed leaders who always find a way — or make one. ENTJs thrive in roles of authority and direction.",
        "strengths": "Leadership, confidence, efficiency",
        "growth": "Patience and emotional sensitivity toward others",
    },
    "ENTP": {
        "title": "The Debater",
        "summary": "Smart, curious thinkers who cannot resist an intellectual challenge. ENTPs love to debate and explore all angles of an idea.",
        "strengths": "Wit, versatility, enthusiasm for ideas",
        "growth": "Focusing energy and respecting boundaries",
    },
    "INFJ": {
        "title": "The Advocate",
        "summary": "Quiet, insightful, and deeply empathetic. INFJs are driven by a core sense of idealism and the desire to help others find their way.",
        "strengths": "Empathy, vision, integrity",
        "growth": "Setting boundaries and avoiding burnout",
    },
    "INFP": {
        "title": "The Mediator",
        "summary": "Poetic and kind-hearted, INFPs are guided by their values and a deep desire for authenticity and meaning in everything they do.",
        "strengths": "Compassion, creativity, open-mindedness",
        "growth": "Managing self-criticism and staying grounded",
    },
    "ENFJ": {
        "title": "The Protagonist",
        "summary": "Charismatic and inspiring leaders who radiate warmth. ENFJs are natural mentors who genuinely care about the growth of those around them.",
        "strengths": "Empathy, communication, altruism",
        "growth": "Prioritising their own needs without guilt",
    },
    "ENFP": {
        "title": "The Campaigner",
        "summary": "Enthusiastic, creative, and sociable free spirits who find reasons to smile everywhere. ENFPs love connecting ideas and people.",
        "strengths": "Energy, empathy, creativity",
        "growth": "Maintaining focus and following through",
    },
    "ISTJ": {
        "title": "The Logistician",
        "summary": "Practical and fact-minded, ISTJs are reliable and dedicated individuals who take responsibilities seriously.",
        "strengths": "Dependability, thoroughness, honesty",
        "growth": "Embracing change and emotional expression",
    },
    "ISFJ": {
        "title": "The Defender",
        "summary": "Warm, caring protectors who are always ready to defend those they love. ISFJs combine practicality with deep emotional loyalty.",
        "strengths": "Reliability, patience, generosity",
        "growth": "Asserting personal needs and resisting overcommitment",
    },
    "ESTJ": {
        "title": "The Executive",
        "summary": "Excellent administrators who are unsurpassed at managing things and people. ESTJs value order, tradition, and clear expectations.",
        "strengths": "Organisation, dedication, directness",
        "growth": "Flexibility and emotional openness",
    },
    "ESFJ": {
        "title": "The Consul",
        "summary": "Caring, social, and popular people who are always eager to help. ESFJs thrive when they feel appreciated and connected to a community.",
        "strengths": "Warmth, practicality, loyalty",
        "growth": "Building self-worth beyond others' opinions",
    },
    "ISTP": {
        "title": "The Virtuoso",
        "summary": "Bold, practical experimenters who love to explore with their hands and eyes. ISTPs master tools and skills with quiet intensity.",
        "strengths": "Problem-solving, calm under pressure, adaptability",
        "growth": "Long-term planning and emotional communication",
    },
    "ISFP": {
        "title": "The Adventurer",
        "summary": "Flexible, charming artists who are always ready to explore something new. ISFPs live in the present and delight in creativity.",
        "strengths": "Creativity, empathy, adaptability",
        "growth": "Confidence and planning for the future",
    },
    "ESTP": {
        "title": "The Entrepreneur",
        "summary": "Smart, energetic, and perceptive people who enjoy living on the edge. ESTPs bring boldness and practicality to every situation.",
        "strengths": "Boldness, directness, perceptiveness",
        "growth": "Patience and consideration of long-term consequences",
    },
    "ESFP": {
        "title": "The Entertainer",
        "summary": "Spontaneous, energetic, and enthusiastic performers who love the spotlight and live for the moment.",
        "strengths": "Optimism, practicality, social energy",
        "growth": "Focus and long-term planning",
    },
}


def get_personality_description(personality_type):
    """Return the description dict for the given type, or a generic fallback."""
    return PERSONALITY_DESCRIPTIONS.get(personality_type.upper() if personality_type else "", {
        "title": "Unique Profile",
        "summary": "Your personality profile is one of a kind. Keep exploring what makes you, you.",
        "strengths": "Self-awareness and openness",
        "growth": "Continued self-exploration",
    })


def generate_mood_chart(user_id):

    from datetime import datetime
    import matplotlib.pyplot as plt

    data = get_mood_data(user_id)

    if not data:
        return None

    data = data[-7:]

    dates = [
        datetime.strptime(d["date"], "%Y-%m-%d").strftime("%b %d")
        for d in data
    ]

    scores = [d["score"] for d in data]

    filepath = f"mood_chart_{user_id}.png"

    plt.figure(figsize=(10, 5))
    plt.style.use('default')

    plt.plot(dates, scores, marker='o', linewidth=2.5, color='#4A90E2')
    plt.fill_between(dates, scores, color='#4A90E2', alpha=0.15)
    plt.scatter(dates[-1], scores[-1], color='red', s=80, label='Today', zorder=5)

    plt.title("Weekly Mood Trend", fontsize=14, fontweight='bold')
    plt.yticks(
        [0, 0.25, 0.5, 0.75, 1],
        ["Very Low", "Low", "Neutral", "Positive", "Very Positive"],
        fontsize=9
    )
    plt.ylim(0, 1)
    plt.xticks(rotation=30, fontsize=9)
    plt.grid(True, linestyle='--', alpha=0.3)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().spines['right'].set_visible(False)
    plt.legend()
    plt.tight_layout()
    plt.savefig(filepath, dpi=300, bbox_inches='tight')
    plt.close()

    return filepath


def get_mood_by_date(user_id, date):
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT score FROM mood_logs
        WHERE user_id = ? AND DATE(created_at) = ?
    """, (user_id, date))
    rows = cursor.fetchall()
    conn.close()
    return [{"score": row[0]} for row in rows]

def get_journal_by_date(user_id, date):
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT content FROM journal
        WHERE user_id = ? AND DATE(created_at) = ?
    """, (user_id, date))
    rows = cursor.fetchall()
    conn.close()
    return rows

def update_daily_summary(user_id):
    from datetime import date
    today = str(date.today())
    conn = get_connection()
    cursor = conn.cursor()

    moods = get_mood_by_date(user_id, today)
    avg_mood = sum([m["score"] for m in moods]) / len(moods) if moods else 0

    journals = get_journal_by_date(user_id, today)
    journal_count = len(journals)

    cursor.execute("""
        SELECT SUM(duration) FROM meditation_logs
        WHERE user_id = ? AND date = ?
    """, (user_id, today))
    meditation = cursor.fetchone()[0] or 0

    if avg_mood > 0.5:
        summary = "Mostly positive day 😊"
    elif avg_mood < -0.5:
        summary = "Tough day 😔"
    else:
        summary = "Balanced day 😐"

    cursor.execute("""
        INSERT OR REPLACE INTO daily_summary (user_id, date, avg_mood, journal_count, summary, meditation)
        VALUES (?, ?, ?, ?, ?, ?)
    """, (user_id, today, avg_mood, journal_count, summary, meditation))

    conn.commit()
    conn.close()


def get_personality(user_id):
    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("""
        SELECT personality_type
        FROM personality_results
        WHERE user_id = ?
        ORDER BY id DESC
        LIMIT 1
    """, (user_id,))

    row = cursor.fetchone()
    conn.close()

    return row[0] if row else "Unknown"


def get_mental_score(user_id):

    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("""
        SELECT score
        FROM mental_health_results
        WHERE user_id = ?
        ORDER BY timestamp DESC
        LIMIT 1
    """, (user_id,))

    row = cursor.fetchone()
    conn.close()

    if not row:
        return None

    raw_score = row[0]
    wellness_score = max(0, 100 - (raw_score * 2))
    return wellness_score

def analyze_mood(user_id):

    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("""
        SELECT score, DATE(created_at)
        FROM mood_logs
        WHERE user_id = ?
    """, (user_id,))

    rows = cursor.fetchall()
    conn.close()

    if not rows:
        return None
    scores = [r[0] for r in rows]

    avg = sum(scores) / len(scores)

    best = max(rows, key=lambda x: x[0])
    worst = min(rows, key=lambda x: x[0])

    return {
        "avg": round(avg, 2),
        "best_day": best[1],
        "best_score": best[0],
        "worst_day": worst[1],
        "worst_score": worst[0]
    }

def mental_interpretation(score):

    if score >= 80:
        return "Your emotional wellbeing appears strong and stable."
    elif score >= 60:
        return "Your mental wellbeing appears generally positive with some fluctuations."
    elif score >= 40:
        return "Some emotional stress may be present. Taking time for self-care may help."
    elif score >= 25:
        return "You may be experiencing emotional strain. Support strategies could be beneficial."
    else:
        return "Emotional distress patterns detected. Speaking with supportive people or professionals may help."
    


def generate_wellness_report(user_id, filepath):

    styles = getSampleStyleSheet()

    personality = get_personality(user_id)
    personality_desc = get_personality_description(personality)
    mental_score = get_mental_score(user_id)
    mood = analyze_mood(user_id)
    
    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT COUNT(*) FROM journal WHERE user_id=?", (user_id,))
    total_journals = cursor.fetchone()[0]

    cursor.execute("SELECT SUM(duration) FROM meditation_logs WHERE user_id=?", (user_id,))
    total_meditation = cursor.fetchone()[0] or 0

    conn.close()

    streak = calculate_streak(user_id)

    elements = []

    elements.append(Paragraph("MindMate Personal Wellness Report", styles["Title"]))
    elements.append(Spacer(1, 20))

    elements.append(Paragraph(f"Generated: {datetime.now().strftime('%Y-%m-%d')}", styles["Normal"]))
    elements.append(Spacer(1, 20))

    # ── Personality Profile ──────────────────────────────────────────────────
    elements.append(Paragraph("<b>Personality Profile</b>", styles["Heading2"]))
    elements.append(Paragraph(
        f"Personality Type: <b>{personality}</b> — {personality_desc['title']}",
        styles["Normal"]
    ))
    elements.append(Spacer(1, 6))
    elements.append(Paragraph(personality_desc["summary"], styles["Normal"]))
    elements.append(Spacer(1, 6))
    elements.append(Paragraph(
        f"<b>Key Strengths:</b> {personality_desc['strengths']}",
        styles["Normal"]
    ))
    elements.append(Paragraph(
        f"<b>Growth Area:</b> {personality_desc['growth']}",
        styles["Normal"]
    ))
    elements.append(Spacer(1, 15))

    # ── Mental Health Capacity ───────────────────────────────────────────────
    elements.append(Paragraph("<b>Mental Health Capacity</b>", styles["Heading2"]))
    elements.append(Paragraph(f"Score: {mental_score} / 100", styles["Normal"]))
    elements.append(Paragraph(mental_interpretation(mental_score), styles["Normal"]))
    elements.append(Spacer(1, 15))

    # ── Mood Analysis ────────────────────────────────────────────────────────
    elements.append(Paragraph("<b>Mood Analysis</b>", styles["Heading2"]))
    elements.append(Paragraph(f"Average Mood Score: {mood['avg']}", styles["Normal"]))
    elements.append(Paragraph(f"Total Journals: {total_journals}", styles["Normal"]))
    elements.append(Paragraph(f"Total Meditation Sessions: {total_meditation} Minutes", styles["Normal"]))
    elements.append(Paragraph(f"Current Streak: {streak} days", styles["Normal"]))
    elements.append(Paragraph(f"Best Mood Day: {mood['best_day']}", styles["Normal"]))
    elements.append(Paragraph(f"Lowest Mood Day: {mood['worst_day']}", styles["Normal"]))

    elements.append(Spacer(1, 15))

    # ── Mood chart ───────────────────────────────────────────────────────────
    chart = generate_mood_chart(user_id)

    if chart:
        from reportlab.platypus import Image
        elements.append(Spacer(1, 20))
        elements.append(Image(chart, width=450, height=250))

    elements.append(Spacer(1, 15))
    elements.append(PageBreak())

    # ── AI Insights ──────────────────────────────────────────────────────────
    insight = generate_ai_insight(mood["avg"], mental_score)

    elements.append(Paragraph("AI Insights", styles["Heading2"]))
    elements.append(Paragraph(insight, styles["Normal"]))
    elements.append(Spacer(1, 15))

    # ── Recommendations ──────────────────────────────────────────────────────
    recommendations = generate_ai_recommendations(mood["avg"], mental_score)

    elements.append(Paragraph("<b>Recommendations</b>", styles["Heading2"]))

    for rec in recommendations:
        elements.append(Paragraph(f"• {rec}", styles["Normal"]))

    elements.append(Spacer(1, 15))

    # ── Overall Summary ──────────────────────────────────────────────────────
    elements.append(Paragraph("<b>Overall Summary</b>", styles["Heading2"]))
    elements.append(Paragraph(
        "This report summarises emotional patterns detected through your interactions with MindMate. "
        "Continued self-awareness and healthy coping strategies can support long-term wellbeing.",
        styles["Normal"]
    ))

    doc = SimpleDocTemplate(filepath, pagesize=A4)
    doc.build(elements)

    if chart and os.path.exists(chart):
        os.remove(chart)

    return filepath


def generate_recommendations(avg_mood, mental_score):

    general = [
        "Maintain a regular sleep schedule to support emotional balance.",
        "Spending time outdoors can help improve mood and reduce stress.",
        "Talking with trusted friends or family members may provide emotional relief.",
        "Taking short breaks during work or study sessions can help prevent burnout.",
        "Engaging in hobbies you enjoy can support positive emotional wellbeing."
    ]

    stress = [
        "Practice breathing exercises or mindfulness techniques to manage stress.",
        "Breaking large tasks into smaller steps may make challenges feel more manageable.",
        "Limiting screen time before sleep may help improve rest quality."
    ]

    low_mood = [
        "Consider journaling your thoughts to better understand emotional patterns.",
        "Gentle physical activity such as walking can help improve mood.",
        "Listening to music or engaging in creative activities may help lift mood."
    ]

    positive = [
        "Continue maintaining the habits that support your positive emotional wellbeing.",
        "Sharing positive experiences with others can strengthen emotional resilience.",
        "Reflecting on accomplishments may help reinforce positive mindset."
    ]

    recommendations = []
    recommendations.extend(random.sample(general, 2))

    if mental_score < 50:
        recommendations.extend(random.sample(low_mood, 2))
    elif avg_mood < 0:
        recommendations.extend(random.sample(stress, 2))
    else:
        recommendations.extend(random.sample(positive, 2))

    return recommendations


from datetime import datetime, timedelta

def calculate_streak(user_id):

    conn = get_connection() 
    cursor = conn.cursor()

    cursor.execute("""
        SELECT DISTINCT date
        FROM daily_summary
        WHERE user_id=?
        ORDER BY date DESC
    """, (user_id,))

    rows = cursor.fetchall()

    if not rows:
        return 0

    streak = 0
    today = datetime.today().date()

    for row in rows:
        entry_date = datetime.strptime(row[0], "%Y-%m-%d").date()
        if entry_date == today - timedelta(days=streak):
            streak += 1
        else:
            break

    return streak

def is_new_user(user_id):
    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT COUNT(*) FROM mood_logs WHERE user_id=?", (user_id,))
    mood_count = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM journal WHERE user_id=?", (user_id,))
    journal_count = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM mental_health_results WHERE user_id=?", (user_id,))
    mental_count = cursor.fetchone()[0]

    conn.close()

    return (mood_count + journal_count + mental_count) == 0


import requests

def generate_ai_recommendations(avg_mood, mental_score):

    prompt = f"""
You are a mental wellness assistant.

User data:
- Average mood: {avg_mood}
- Mental health score: {mental_score}/100

Give 3 short, practical, supportive recommendations.

Rules:
- Keep each recommendation 1 sentence
- Be human and empathetic
- Do NOT repeat phrases
- No numbering
"""

    try:
        res = requests.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers={
                "Authorization": f"Bearer {GROQ_API_KEY}",
                "Content-Type": "application/json"
            },
            json={
                "model": "llama-3.3-70b-versatile",
                "messages": [
                    {"role": "system", "content": "You help with mental wellness."},
                    {"role": "user", "content": prompt}
                ],
                "temperature": 0.7
            }
        )

        data = res.json()
        text = data["choices"][0]["message"]["content"]
        recommendations = [line.strip("-• ") for line in text.split("\n") if line.strip()]
        return recommendations[:3]

    except Exception as e:
        print("AI recommendation error:", e)
        return [
            "Take a short break and breathe slowly.",
            "Talk to someone you trust.",
            "Get some fresh air or light movement."
        ]


def generate_ai_insight(avg_mood, mental_score):
    prompt = f"""
Analyse this user's mental wellness:

- Avg mood: {avg_mood}

Write a short 2-3 line insight about their emotional pattern.
Be supportive and human.
"""

    try:
        res = requests.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers={
                "Authorization": f"Bearer {GROQ_API_KEY}",
                "Content-Type": "application/json"
            },
            json={
                "model": "llama-3.3-70b-versatile",
                "messages": [
                    {"role": "user", "content": prompt}
                ]
            }
        )

        return res.json()["choices"][0]["message"]["content"]

    except:
        return "Your emotional patterns show balance with areas to improve. Keep taking care of yourself."

File: frontend/src/components/report.js
import { useEffect, useState } from "react";
import ReportMoodChart from "./ReportMoodChart";
import API_URL from "../config";
import useUser from "../hooks/useUser";

// ─────────────────────────────────────────────
// Personality type descriptions (MBTI-style)
// ─────────────────────────────────────────────
const PERSONALITY_DESCRIPTIONS = {
  INTJ: { title: "The Architect",     summary: "Strategic, independent, and highly analytical. INTJs are driven by logic and long-term vision, preferring depth over breadth in all pursuits.",            strengths: "Strategic thinking, decisiveness, high standards",          growth: "Opening up emotionally and embracing flexibility" },
  INTP: { title: "The Logician",      summary: "Curious and inventive thinkers who love exploring abstract ideas. INTPs seek precision and often lose themselves in thought.",                              strengths: "Analytical ability, creativity, open-mindedness",            growth: "Following through on ideas and social engagement" },
  ENTJ: { title: "The Commander",     summary: "Bold, imaginative, and strong-willed leaders who always find a way — or make one. ENTJs thrive in roles of authority and direction.",                     strengths: "Leadership, confidence, efficiency",                         growth: "Patience and emotional sensitivity toward others" },
  ENTP: { title: "The Debater",       summary: "Smart, curious thinkers who cannot resist an intellectual challenge. ENTPs love to debate and explore all angles of an idea.",                            strengths: "Wit, versatility, enthusiasm for ideas",                     growth: "Focusing energy and respecting boundaries" },
  INFJ: { title: "The Advocate",      summary: "Quiet, insightful, and deeply empathetic. INFJs are driven by a core sense of idealism and the desire to help others find their way.",                   strengths: "Empathy, vision, integrity",                                 growth: "Setting boundaries and avoiding burnout" },
  INFP: { title: "The Mediator",      summary: "Poetic and kind-hearted, INFPs are guided by their values and a deep desire for authenticity and meaning in everything they do.",                        strengths: "Compassion, creativity, open-mindedness",                    growth: "Managing self-criticism and staying grounded" },
  ENFJ: { title: "The Protagonist",   summary: "Charismatic and inspiring leaders who radiate warmth. ENFJs are natural mentors who genuinely care about the growth of those around them.",              strengths: "Empathy, communication, altruism",                           growth: "Prioritising their own needs without guilt" },
  ENFP: { title: "The Campaigner",    summary: "Enthusiastic, creative, and sociable free spirits who find reasons to smile everywhere. ENFPs love connecting ideas and people.",                        strengths: "Energy, empathy, creativity",                                growth: "Maintaining focus and following through" },
  ISTJ: { title: "The Logistician",   summary: "Practical and fact-minded, ISTJs are reliable and dedicated individuals who take responsibilities seriously.",                                            strengths: "Dependability, thoroughness, honesty",                       growth: "Embracing change and emotional expression" },
  ISFJ: { title: "The Defender",      summary: "Warm, caring protectors who are always ready to defend those they love. ISFJs combine practicality with deep emotional loyalty.",                        strengths: "Reliability, patience, generosity",                          growth: "Asserting personal needs and resisting overcommitment" },
  ESTJ: { title: "The Executive",     summary: "Excellent administrators who are unsurpassed at managing things and people. ESTJs value order, tradition, and clear expectations.",                      strengths: "Organisation, dedication, directness",                       growth: "Flexibility and emotional openness" },
  ESFJ: { title: "The Consul",        summary: "Caring, social, and popular people who are always eager to help. ESFJs thrive when they feel appreciated and connected to a community.",                strengths: "Warmth, practicality, loyalty",                              growth: "Building self-worth beyond others' opinions" },
  ISTP: { title: "The Virtuoso",      summary: "Bold, practical experimenters who love to explore with their hands and eyes. ISTPs master tools and skills with quiet intensity.",                       strengths: "Problem-solving, calm under pressure, adaptability",         growth: "Long-term planning and emotional communication" },
  ISFP: { title: "The Adventurer",    summary: "Flexible, charming artists who are always ready to explore something new. ISFPs live in the present and delight in creativity.",                        strengths: "Creativity, empathy, adaptability",                          growth: "Confidence and planning for the future" },
  ESTP: { title: "The Entrepreneur",  summary: "Smart, energetic, and perceptive people who enjoy living on the edge. ESTPs bring boldness and practicality to every situation.",                       strengths: "Boldness, directness, perceptiveness",                       growth: "Patience and consideration of long-term consequences" },
  ESFP: { title: "The Entertainer",   summary: "Spontaneous, energetic, and enthusiastic performers who love the spotlight and live for the moment.",                                                    strengths: "Optimism, practicality, social energy",                      growth: "Focus and long-term planning" },
};

function getPersonalityDesc(type) {
  return PERSONALITY_DESCRIPTIONS[type?.toUpperCase()] || {
    title: "Unique Profile",
    summary: "Your personality profile is one of a kind. Keep exploring what makes you, you.",
    strengths: "Self-awareness and openness",
    growth: "Continued self-exploration",
  };
}

function StatCard({ label, value, icon, color }) {
  return (
    <div style={{ background: "#fff", borderRadius: "14px", padding: "18px 20px", border: "1px solid #f1f5f9", boxShadow: "0 2px 10px rgba(0,0,0,0.05)", display: "flex", alignItems: "center", gap: "14px" }}>
      <div style={{ fontSize: "24px", width: "44px", height: "44px", borderRadius: "12px", background: color + "22", display: "flex", alignItems: "center", justifyContent: "center" }}>{icon}</div>
      <div>
        <p style={{ margin: "0 0 2px", fontSize: "0.75rem", fontWeight: "600", color: "#94a3b8", textTransform: "uppercase", letterSpacing: "0.5px" }}>{label}</p>
        <p style={{ margin: 0, fontSize: "1.1rem", fontWeight: "700", color: "#1e293b" }}>{value}</p>
      </div>
    </div>
  );
}

function Section({ title, children }) {
  return (
    <div style={{ background: "#fff", borderRadius: "16px", padding: "24px 28px", boxShadow: "0 2px 12px rgba(0,0,0,0.05)", border: "1px solid #f1f5f9", marginBottom: "16px" }}>
      <h3 style={{ margin: "0 0 16px", fontSize: "1rem", fontWeight: "700", color: "#1e293b", paddingBottom: "12px", borderBottom: "1px solid #f1f5f9" }}>{title}</h3>
      {children}
    </div>
  );
}

function Report() {
  const user = useUser();
  const [report, setReport] = useState(null);
  const [error, setError] = useState("");

  const today = new Date();
  const year = today.getFullYear();
  const month = today.getMonth() + 1;

  useEffect(() => {
    if (!user.id) return;
    fetch(`${API_URL}/report/data/${user.id}`)
      .then(res => res.json())
      .then(data => setReport(data))
      .catch(() => setError("Failed to load report. Please try again."));
  }, [user.id]);

  if (error) return (
    <div style={{ padding: "20px" }}>
      <div style={{ background: "#fef2f2", border: "1px solid #fecaca", borderRadius: "12px", padding: "16px 20px", color: "#dc2626" }}>⚠️ {error}</div>
    </div>
  );

  if (!report) return (
    <div style={{ padding: "40px", textAlign: "center", color: "#94a3b8" }}>
      <div style={{ fontSize: "36px", marginBottom: "12px" }}>📊</div>
      <p>Loading your wellness report...</p>
    </div>
  );

  if (report.empty) return (
    <div style={{ maxWidth: "500px", margin: "60px auto", textAlign: "center" }}>
      <div style={{ fontSize: "48px", marginBottom: "16px" }}>💙</div>
      <h2 style={{ color: "#1e293b", marginBottom: "8px" }}>Welcome to MindMate</h2>
      <p style={{ color: "#64748b" }}>No data yet. Start chatting, journaling, or meditating to generate your wellness report.</p>
    </div>
  );

  const getMoodColor = (score) => score >= 1 ? "#22c55e" : score >= 0 ? "#eab308" : "#ef4444";
  const personalityDesc = getPersonalityDesc(report.personality);

  return (
    <div style={{ maxWidth: "900px", margin: "0 auto", padding: "8px 0 32px" }}>

      {/* Header */}
      <div style={{ background: "linear-gradient(135deg, #1e3a8a 0%, #2563eb 60%, #60a5fa 100%)", borderRadius: "20px", padding: "28px 32px", marginBottom: "24px", color: "white", display: "flex", justifyContent: "space-between", alignItems: "flex-start" }}>
        <div>
          <h1 style={{ margin: "0 0 6px", fontSize: "1.5rem", fontWeight: "700" }}>📄 Wellness Report</h1>
          <p style={{ margin: 0, color: "rgba(255,255,255,0.75)", fontSize: "0.9rem" }}>
            {today.toLocaleDateString("en-IN", { day: "numeric", month: "long", year: "numeric" })}
          </p>
        </div>
        <button
          onClick={() => window.open(`${API_URL}/report/${user.id || 1}`)}
          style={{ padding: "10px 20px", border: "2px solid rgba(255,255,255,0.4)", borderRadius: "10px", background: "rgba(255,255,255,0.15)", color: "white", fontWeight: "600", fontSize: "0.85rem", cursor: "pointer", transition: "all 0.2s" }}
          onMouseEnter={e => e.target.style.background = "rgba(255,255,255,0.25)"}
          onMouseLeave={e => e.target.style.background = "rgba(255,255,255,0.15)"}
        >
          ⬇️ Download PDF
        </button>
      </div>

      {/* Stats row */}
      <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fill, minmax(180px, 1fr))", gap: "12px", marginBottom: "20px" }}>
        <StatCard label="Streak" value={`${report.streak} days`} icon="🔥" color="#f97316" />
        <StatCard label="Journals" value={report.total_journals} icon="📓" color="#2563eb" />
        <StatCard label="Meditation" value={`${report.total_meditation} mins`} icon="🧘" color="#10b981" />
        <StatCard label="Avg Mood" value={report.avg_mood ?? "—"} icon="📊" color={getMoodColor(report.avg_mood)} />
      </div>

      {/* Personality */}
      <Section title="🧠 Personality Profile">
        {/* Type badge row */}
        <div style={{ display: "flex", alignItems: "center", gap: "16px", marginBottom: "16px" }}>
          <div style={{ fontSize: "2.2rem", fontWeight: "800", color: "#1e40af", background: "#eff6ff", border: "2px solid #bfdbfe", borderRadius: "14px", padding: "10px 20px", letterSpacing: "2px", flexShrink: 0 }}>
            {report.personality || "—"}
          </div>
          <div>
            <p style={{ margin: "0 0 2px", fontWeight: "700", color: "#1e293b", fontSize: "1rem" }}>{personalityDesc.title}</p>
            <p style={{ margin: 0, color: "#64748b", fontSize: "0.88rem" }}>Based on your personality assessment</p>
          </div>
        </div>

        {/* Description card */}
        <div style={{ background: "#f8fafc", borderRadius: "12px", padding: "16px 18px", border: "1px solid #e2e8f0" }}>
          <p style={{ margin: "0 0 12px", color: "#374151", fontSize: "0.92rem", lineHeight: 1.7 }}>
            {personalityDesc.summary}
          </p>
          <div style={{ display: "grid", gridTemplateColumns: "1fr 1fr", gap: "10px" }}>
            <div style={{ background: "#f0fdf4", borderRadius: "10px", padding: "10px 14px", border: "1px solid #bbf7d0" }}>
              <p style={{ margin: "0 0 4px", fontSize: "0.72rem", fontWeight: "700", color: "#16a34a", textTransform: "uppercase", letterSpacing: "0.5px" }}>✦ Key Strengths</p>
              <p style={{ margin: 0, fontSize: "0.87rem", color: "#1e293b" }}>{personalityDesc.strengths}</p>
            </div>
            <div style={{ background: "#fefce8", borderRadius: "10px", padding: "10px 14px", border: "1px solid #fde68a" }}>
              <p style={{ margin: "0 0 4px", fontSize: "0.72rem", fontWeight: "700", color: "#ca8a04", textTransform: "uppercase", letterSpacing: "0.5px" }}>✦ Growth Area</p>
              <p style={{ margin: 0, fontSize: "0.87rem", color: "#1e293b" }}>{personalityDesc.growth}</p>
            </div>
          </div>
        </div>
      </Section>

      {/* Mental Health */}
      <Section title="🧪 Mental Health Score">
        <div style={{ display: "flex", alignItems: "center", gap: "20px", flexWrap: "wrap" }}>
          <div style={{ position: "relative", width: "80px", height: "80px", flexShrink: 0 }}>
            <svg viewBox="0 0 36 36" style={{ width: "80px", height: "80px", transform: "rotate(-90deg)" }}>
              <circle cx="18" cy="18" r="15.9" fill="none" stroke="#f1f5f9" strokeWidth="3" />
              <circle cx="18" cy="18" r="15.9" fill="none" stroke={getMoodColor((report.mental_score || 0) / 50 - 1)}
                strokeWidth="3" strokeDasharray={`${report.mental_score || 0} 100`} strokeLinecap="round" />
            </svg>
            <div style={{ position: "absolute", inset: 0, display: "flex", alignItems: "center", justifyContent: "center", fontSize: "1rem", fontWeight: "700", color: "#1e293b" }}>
              {report.mental_score ?? "—"}
            </div>
          </div>
          <div>
            <p style={{ margin: "0 0 4px", fontWeight: "600", color: "#1e293b" }}>Score: {report.mental_score} / 100</p>
            <p style={{ margin: 0, color: "#64748b", fontSize: "0.88rem", lineHeight: 1.6 }}>{report.mental_interpretation}</p>
          </div>
        </div>
      </Section>

      {/* Mood Analysis */}
      <Section title="📊 Mood Analysis">
        <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fill, minmax(200px, 1fr))", gap: "10px", marginBottom: "16px" }}>
          {[
            ["Best Day", report.best_day, "😊"],
            ["Lowest Day", report.worst_day, "😔"],
            ["Best Score", report.best_score, "⬆️"],
            ["Lowest Score", report.worst_score, "⬇️"],
          ].map(([label, val, icon]) => (
            <div key={label} style={{ background: "#f8fafc", borderRadius: "10px", padding: "12px 14px" }}>
              <p style={{ margin: "0 0 2px", fontSize: "0.75rem", color: "#94a3b8", fontWeight: "600" }}>{icon} {label}</p>
              <p style={{ margin: 0, fontWeight: "700", color: "#1e293b" }}>{val ?? "—"}</p>
            </div>
          ))}
        </div>
      </Section>

      {/* Mood chart */}
      <Section title="📈 Weekly Mood Trend">
        <div style={{ height: "260px" }}>
          <ReportMoodChart moodData={report.weekly_mood} year={year} month={month} />
        </div>
      </Section>

      {/* AI Insights */}
      <Section title="🤖 AI Insights">
        <p style={{ margin: 0, color: "#374151", fontSize: "0.92rem", lineHeight: 1.8, background: "#f8fafc", borderRadius: "10px", padding: "14px 16px" }}>{report.insight}</p>
      </Section>

      {/* Recommendations */}
      <Section title="💡 Recommendations">
        {report.recommendations?.length > 0 ? (
          <div style={{ display: "flex", flexDirection: "column", gap: "10px" }}>
            {report.recommendations.map((rec, i) => (
              <div key={i} style={{ display: "flex", alignItems: "flex-start", gap: "10px", padding: "12px 14px", background: "#f0fdf4", borderRadius: "10px", border: "1px solid #bbf7d0" }}>
                <span style={{ color: "#22c55e", fontWeight: "700", flexShrink: 0 }}>✓</span>
                <span style={{ fontSize: "0.9rem", color: "#374151", lineHeight: 1.6 }}>{rec}</span>
              </div>
            ))}
          </div>
        ) : (
          <p style={{ color: "#94a3b8", margin: 0 }}>No recommendations available yet.</p>
        )}
      </Section>
    </div>
  );
}

export default Report;